### ISSUE TITLE
Running bionix under a cluster

### ISSUE THREAD

**Issue by pachiras** [PapenfussLab/bionix#18]
Hi @jbedo,

I am testing bionix and I really like this.
Controlling environment, software and workflow by some simple scripts 
with a single language is fantastic.

But I run into a problem in the tests.

We have four computing nodes and four file servers.

- Computing Node ( 1Tb SSD) x4
- File Servers (168Tb HDD) x4

I had installed `nix` one of the computing nodes.
We usually use `slurm` to control jobs. But for bionix I didn't use it since it was a test.

When the bionix finishes its calculation, I found that all the calculation results
were _installed_ in the nix store and the files were not accessible from other nodes.
And this configuration consumes the system partition on the SSD very fast.

I know bionix is still under development, but what do you think is the best practice 
for the cluster environment like this?

Thanks.

**Comment by jbedo** [comment]
Thanks for the comments :)

I'm a little unclear how your cluster is configured, it sounds like it's managed by slurm but doesn't have shared storage? I've never really thought about this a particular configuration and it is probably quite challenging. Our slurm cluster has shared storage and we just put the nix store on that so that all nodes can can access it during builds, and Nix handles the job submissions by calling salloc/srun on a node with job submissions privileges. Should you wish to do that, there is a slurm wrapper implementation in the bionix tree and an example that configures the slurm submissions. I've a newer example that's simpler that uses overlays to do the same thing that I'll make a priority to clean up and publish next week.

Otherwise, if you have a cluster with non-shared storage then Nix has distributed build features available (see https://nixos.wiki/wiki/Distributed_build and https://nixos.org/nix/manual/#chap-distributed-builds). The basic idea is to setup SSH access between the remotes, along with some signing keys. The remotes can then execute builds and handling copying of things between the stores for you. In this case, though, slurm doesn't really neatly fit in.

As for the store growing fast, this is unavoidable if dealing with large datasets, but most of it is garbage and will therefore be reclaimed when garbage collection executes, either by running `nix-collect-garbage` or when it is invoked automatically. For the latter, you can configure GC to run automatically when free space reduces to a certain point (see min-free and max-free in the config https://nixos.org/manual/nix/stable/#description-42 and the wiki https://nixos.wiki/wiki/Storage_optimization#cite_note-4).

**Comment by pachiras** [comment]
Thank you for the detailed advice.
And sorry I was short of explanations.

Each calculation server mounts every file server storage using NFS:
```
  node1: /data/fileserver1
              /data/fileserver2
              /data/fileserver3
              /data/fileserver4

  node2: /data/fileserver1
              /data/fileserver2
              /data/fileserver3
              /data/fileserver4
.. and so on
````

And here is what I did:

1. Install `nix` on `node1`.
2. Log in to node1. Change directory to `/data/fileserver1/workspace`, which is accessible from all calculation nodes.
3. Start bionix from that directory. Check the result `/data/fileserver1/workspace/result`.
4. Log out from `node1` and Log in to `node2`. Access the file `/data/fileserver1/workspace/result`.

At this point I found the `result` file is just a symbolic link pointing to `/nix` on `node1` where users in `node2` cannot access.

> Our slurm cluster has shared storage and we just put the nix store on that so that all nodes can can access it during builds

That's it! I was wrongly believing that `nix` had to be installed on each calculation node. After sharing `/nix` on `node1`, my accession problem was solved. Thank you.

---
As for the system disk consumption problem, I still, however, have to think about it, because some projects easily consume more than 1Tb disk capacity and stall the system. Although mounting a big disk on `/nix` and monitoring the disk usage is one of the solutions, but I'm wondering if there is a more simple solution.

Ideally speaking, I wish I had a command like this  (I know this is not so easy to implement though):

1. We can make a command `bionix`
2. `bionix` starts calculation by typing `bionix build`
3. `bionix` creates _a local store_ where command was invoked.
4. `bionix` uses both `/nix/store` and  _local store_. `/nix/store` is for software packages and the _local store_ is for analysis result and intermediate files.

The local store would look like this

```
my_workspace
├── project1
│   ├── bionix
│   │   └── store
│   │       ├── z46d9zygv7gg3cmx17ldyan13vwbrfwp-result
│   │       └── zv6i9ika2x4m8krvbg8c4xxr0cll6ang-sample1-1.fq.drv
│   └── result -> bionix/store/z46d9zygv7gg3cmx17ldyan13vwbrfwp-result
└── project2
    ├── bionix
    │   └── store
    │       ├── z2fnp3sqvfl7pjwkf305230i4f3s325d-sample2-1.fq.drv
    │       └── z3vrwybyn961scqcfj8cvb4mjkp93cyx-result
    └── result -> bionix/store/z3vrwybyn961scqcfj8cvb4mjkp93cyx-result
```

If you don’t need `project1` any more, you can simply delete the directory `project1`.
The other day I found that `cabal` implements [nix-style build system] (https://cabal.readthedocs.io/en/3.4/nix-local-build-overview.html). I am wondering if the same kind of local store system can be implemented in bionix :thinking:

**Comment by jbedo** [comment]
You have some great ideas about layout there, and it's somewhat close to how I've structured things for one of my projects. In our case we have organised things so that each user has their own store located on fast scratch storage when executing builds, with the results then copied to another store located on slower tape-backed project storage. This is similar to your proposal with the additional temporary user store on fast scratch. At the moment I haven't got a `bionix` tool to do this, it's a collection of shell scripts that's pretty site specific, but I agree that if we work out some standard layouts and operating procedures then some tooling to handle it would be ideal.

I had not realised cabal had a new nix-style build engine, thanks for pointing that out. I don't think we need to go to that extent though, we should be able to use Nix with some manipulation of the namespaces to place things where we want, at least that's how my scripts currently work. There's some gotchas though and careful use of side effects is needed to make sure build products are sharable between all the stores.
